### PR TITLE
[#155934635, #160198508] Fix push notification interaction and deep link navigation

### PR DESCRIPTION
**Pull request by DenisFrezzato**
This PR applies a patch to `react-native-push-notification` (making leverage on `patch-package`). I got the code from this answer: https://github.com/zo0r/react-native-push-notification/issues/547#issuecomment-351713972. This https://github.com/zo0r/react-native-push-notification/pull/793 should fix the issue, but I couldn't understand if it has been shipped with last release, which fail in build time (see https://github.com/zo0r/react-native-push-notification/issues/808).

This is obviously a temporary fix.

### CODE REVIEW

File: patches/react-native+0.55.4.patch [hunk 1]
```diff
@@ -1,16 +0,0 @@
---- a/node_modules/react-native/Libraries/Text/TextInput/RCTBaseTextInputView.m
```
Comment (cloudify) [inline patches/react-native+0.55.4.patch]: you don't want to remove this one :)

**Comment by digitalcitizenship** [comment]
## Affected stories

  * 🌟 [#155934635](https://www.pivotaltracker.com/story/show/155934635): Come CIT, alla ricezione di una notifica se clicco sulla notifica voglio visualizzare il dettaglio del messaggio


New dependencies added: <a href='https://www.npmjs.com/package/postinstall-prepare'><code>postinstall-prepare</code></a>.


<h2><a href="https://github.com/ds300/patch-package#readme">postinstall-prepare</a></h2>
<p>Author: David Sheldrick</p>
<p>Description: postinstall-prepare ===================</p>
<p>Homepage: <a href="https://github.com/ds300/patch-package#readme">https://github.com/ds300/patch-package#readme</a></p>

<table>
  <thead><tr><th></th><th width="100%"></th></tr></thead>
  <tr><td>Created</td><td>8 months ago</td></tr><tr><td>Last Updated</td><td>8 months ago</td></tr><tr><td>License</td><td>MIT</td></tr><tr><td>Maintainers</td><td>1</td></tr><tr><td>Releases</td><td>2</td></tr>
</table>

<details>
<summary><code>README</code></summary></br>

postinstall-prepare
===================

Run your app's `prepare` npm script during this package's `postinstall` script.

## Why

Yarn runs the `prepare` script after `yarn`, `yarn install` and `yarn add <package>` but not after `yarn remove <package>`. If you add this package to your project, it will execute your project's `prepare` hook even after a `yarn remove <package>`. This requires your `prepare` script to be idempotent, becase it will be run twice for `yarn`, `yarn install`, and `yarn add <package>`


</details>


<p align="right">
  Generated by :no_entry_sign: <a href="http://github.com/danger/danger-js/">dangerJS</a>
</p>

File: package.json [hunk 2]
```diff
@@ -94,6 +94,7 @@
     "jest": "23.4.1",
     "node-fetch": "^2.1.2",
     "patch-package": "^5.1.1",
+    "postinstall-prepare": "^1.0.1",
```
Comment (cloudify) [inline package.json]: why this?

Comment (DenisFrezzato) [reply package.json]: https://www.npmjs.com/package/patch-package#why-use-postinstall-prepare-with-yarn

Comment (cloudify) [reply package.json]: yes I understand what it does, I mean why you need it to fix this particular issue with push notifications

Comment (DenisFrezzato) [reply package.json]: Because I used `patch-package` to create the patch. `postinstall-prepare` is required by `patch-package` when using `parn` in order to apply the patch.

**Comment by DenisFrezzato** [comment]
This PR is now WIP. When the application is not in foreground, the handling of the push notification no longer works because of https://github.com/teamdigitale/italia-app/pull/396.

**Comment by DenisFrezzato** [comment]
I created a _deferred actions_ system, in order to handle actions which has to be dispatched once the user is logged in, whatever the action.

File: ts/sagas/startup/saveNavigationStateSaga.ts [hunk 3]
```diff
@@ -1,17 +1,21 @@
...
 
 /**
- * Saves the navigation state in the deep link state so that when the app
- * goes through the initialization saga, the user gets sent back to the saved
- * navigation route.
- * Saving and restoring routes relies on the deep link mechanism.
+ * Saves the navigation state in a deferred navigation action so that when
+ * the app  goes through the initialization saga, the user gets sent back
+ * to the saved  navigation route.
+ * Saving and restoring routes relies on the deferred actions mechanism.
```
Comment (cloudify) [inline ts/sagas/startup/saveNavigationStateSaga.ts]: It's worth describing this mechanism in the README and link to it from here.

File: ts/sagas/startup/saveNavigationStateSaga.ts [hunk 4]
```diff
@@ -1,17 +1,21 @@
...
+import { deferToLogin } from "../../store/actions/deferred";
 import { navigationStateSelector } from "../../store/reducers/navigation";
 
 /**
- * Saves the navigation state in the deep link state so that when the app
- * goes through the initialization saga, the user gets sent back to the saved
- * navigation route.
- * Saving and restoring routes relies on the deep link mechanism.
+ * Saves the navigation state in a deferred navigation action so that when
+ * the app  goes through the initialization saga, the user gets sent back
```
Comment (cloudify) [inline ts/sagas/startup/saveNavigationStateSaga.ts]: initialization saga -> startup saga

File: ts/sagas/watchDeferredActionsSaga.ts [hunk 5]
```diff
@@ -0,0 +1,20 @@
+import { Effect, put, select, takeEvery } from "redux-saga/effects";
+import { DEFER_TO_LOGIN } from "../store/actions/constants";
+import { DeferToLogin, pushToDeferredActions } from "../store/actions/deferred";
+import { isPinLoginValidSelector } from "../store/reducers/pinlogin";
+
+function* handleDeferredActionSaga({
```
Comment (cloudify) [inline ts/sagas/watchDeferredActionsSaga.ts]: Please add a comment that explain the purpose of this function

File: ts/sagas/watchDeferredActionsSaga.ts [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+  payload: action
+}: DeferToLogin): IterableIterator<Effect> {
+  const isPinValid: boolean = yield select(isPinLoginValidSelector);
+
+  if (isPinValid) {
+    yield put(action);
```
Comment (cloudify) [inline ts/sagas/watchDeferredActionsSaga.ts]: I don't understand this - very hard to follow - if PIN is valid the action gets triggered right away? if not it gets deferred?

File: ts/sagas/watchNavigateToMessageDetailsSaga.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+import { call, take } from "redux-saga/effects";
+import { GetMessageT, GetServiceT } from "../api/backend";
+import { NAVIGATE_TO_MESSAGE_DETAILS } from "../store/actions/constants";
+import { navigateToMessageDetailsSaga } from "./startup/watchLoadMessagesSaga";
+
+export function* watchNavigateToMessageDetailsSaga(
+  getMessage: TypeofApiCall<GetMessageT>,
+  getService: TypeofApiCall<GetServiceT>
+): IterableIterator<Effect> {
+  while (true) {
```
Comment (cloudify) [inline ts/sagas/watchNavigateToMessageDetailsSaga.ts]: while/take block == takeAll

File: ts/store/actions/deferred.ts [hunk 8]
```diff
@@ -0,0 +1,49 @@
+/**
+ * Action types and action creators related to deferred actions.
+ * A deferred action is an action dispatched before the user login
+ * but actually has to be dispatched once the has been logged in.
```
Comment (cloudify) [inline ts/store/actions/deferred.ts]: ..the user?

File: ts/store/actions/deferred.ts [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+  PUSH_TO_DEFERRED_ACTIONS
+} from "./constants";
+import { Action } from "./types";
+
+export interface DeferToLogin {
+  type: typeof DEFER_TO_LOGIN;
+  payload: Action;
+}
+
+export const deferToLogin = (action: Action): DeferToLogin => ({
```
Comment (cloudify) [inline ts/store/actions/deferred.ts]: I think we need a better name for this action, before reading the explanation above I thought we were talking about deferring to the _next_ login - what about `deferActionAfterLogin`

File: ts/store/actions/deferred.ts [hunk 10]
```diff
@@ -0,0 +1,49 @@
+/**
+ * Action types and action creators related to deferred actions.
+ * A deferred action is an action dispatched before the user login
+ * but actually has to be dispatched once the has been logged in.
+ * (e.g. navigation coming from push notification or from deep link happened
+ * while the app is not in foreground and/or the user is not logged in)
```
Comment (cloudify) [inline ts/store/actions/deferred.ts]: I'm not sure why we need such a generic mechanism for deferring multiple arbitrary actions, aren't we talking just about a single navigation action? what are the use cases for such a generic mechanism?

**Comment by cloudify** [comment]
@DenisFrezzato I'm not really convinced about the purpose of this PR, how does it improve the previous logic?

File: ts/sagas/watchDeferredActionsSaga.ts (hunk 6 above)
Comment (DenisFrezzato) [reply ts/sagas/watchDeferredActionsSaga.ts]: If the pin is valid, then the user is logged in, so the action will be dispatched right away. Otherwise, the action will be dispatched as soon as the user is logged in.

File: ts/store/actions/deferred.ts (hunk 10 above)
Comment (DenisFrezzato) [reply ts/store/actions/deferred.ts]: The use cases, at this time, are the interaction with a push notification (message notification -> navigate to the message, maybe there will be others?) and navigation from a deep link.

File: ts/store/actions/deferred.ts (hunk 9 above)
Comment (DenisFrezzato) [reply ts/store/actions/deferred.ts]: Ok, `deferActionAfterLogin` sounds better.

**Comment by cloudify** [comment]
@DenisFrezzato it makes sense, thanks - but I think this app already has too many moving parts (and things that happen asynchronously that are very hard to reason about) - I'd like to keep the moving parts at the minimum, so unless we really really need it, I'd like to avoid this generic mechanism for deferring actions and stick to a something related to "where the app should navigate to after the login".

File: ts/RootContainer.tsx [hunk 11]
```diff
@@ -38,13 +38,13 @@ class RootContainer extends React.PureComponent<Props> {
...
     if (Platform.OS === "android") {
+      // If the application is summoned from a deep link intent.
       Linking.getInitialURL()
         .then(this.navigateToUrlHandler)
         .catch(console.error); // tslint:disable-line:no-console
-    } else {
-      Linking.addEventListener("url", this.handleOpenUrlEvent);
     }
 
+    Linking.addEventListener("url", this.handleOpenUrlEvent);
```
Comment (cloudify) [inline ts/RootContainer.tsx]: we should probably remove the event listener on `componentWillUnmount`

Comment (cloudify) [reply ts/RootContainer.tsx]: actually, this gets executed also for Android, do we need an `else` ?

**Comment by DenisFrezzato** [comment]
markdown-deep-link

File: ts/RootContainer.tsx (hunk 11 above)
Comment (DenisFrezzato) [reply ts/RootContainer.tsx]: This is the part that fixes the changed behaviour introduced by https://github.com/teamdigitale/italia-app/pull/338. Now, this is the listener for both iOS and Android that handles the deep link navigation event.

File: ts/RootContainer.tsx [hunk 12]
```diff
@@ -45,55 +32,31 @@ class RootContainer extends React.PureComponent<Props> {
...
   public componentWillUnmount() {
-    if (Platform.OS === "ios") {
-      Linking.removeEventListener("url", this.handleOpenUrlEvent);
-    }
-
+    Linking.removeEventListener("url", this.handleOpenUrlEvent);
     AppState.removeEventListener("change", this.handleApplicationActivity);
   }
 
   // public shouldComponentUpdate(_: Props): boolean {
```
Comment (cloudify) [inline ts/RootContainer.tsx]: can you remove this code too?

File: ts/store/reducers/deferred-navigation.ts [hunk 13]
```diff
@@ -0,0 +1,54 @@
+/**
```
Comment (cloudify) [inline ts/store/reducers/deferred-navigation.ts]: use camel case when naming files `deferredNavigation.ts`

File: ts/store/reducers/deferred-navigation.ts [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+  SAVE_NAVIGATION_STATE,
+  SAVE_NAVIGATION_WHEN_LOGGED_IN
+} from "../actions/constants";
+import { Action } from "../actions/types";
+import { GlobalState } from "./types";
+
+export interface DeferredNavigationActionState {
+  // Navigation saved before going to background.
+  navigationState: Action | undefined;
+  navigation: Action | undefined;
```
Comment (cloudify) [inline ts/store/reducers/deferred-navigation.ts]: explain the difference between `navigationState` and `navigation` and what's stored in them

File: ts/store/reducers/index.ts [hunk 15]
```diff
@@ -31,7 +31,7 @@ import walletReducer from "./wallet";
 export const authenticationPersistConfig: PersistConfig = {
   key: "authentication",
   storage: createSecureStorage(),
-  blacklist: ["deepLink"]
+  blacklist: ["deferred-navigation"]
```
Comment (cloudify) [inline ts/store/reducers/index.ts]: is the `deferred-navigation` state stored in the `authentication` store? @fpersico

File: package.json [hunk 16]
```diff
@@ -105,6 +105,7 @@
     "js-yaml": "^3.12.0",
     "node-fetch": "^2.2.0",
     "patch-package": "^5.1.1",
+    "postinstall-prepare": "^1.0.1",
```
Comment (fpersico) [inline package.json]: patch-package da poco consiglia postinstall-postinstall invece che postinstall-prepare

File: patches/react-native-push-notification+3.0.2.patch [hunk 17]
```diff
@@ -0,0 +1,53 @@
+patch-package
+--- a/node_modules/react-native-push-notification/android/src/main/java/com/dieam/reactnativepushnotification/modules/RNPushNotification.java
```
Comment (fpersico) [inline patches/react-native-push-notification+3.0.2.patch]: Questa patch non va bene, viene chiamato due volte onNotification quando si clicca su una notifica mentre l'app è in background.

File: ts/RootContainer.tsx [hunk 18]
```diff
@@ -109,19 +72,12 @@ class RootContainer extends React.PureComponent<Props> {
...
 const mapDispatchToProps = {
   applicationChangeState,
-  setDeepLink,
-  navigateToDeepLink
+  navigateIfLoggedIn
 };
 
 export default connect(
-  mapStateToProps,
+  undefined,
```
Comment (fpersico) [inline ts/RootContainer.tsx]: Why undefined?

File: ts/store/reducers/index.ts (hunk 15 above)
Comment (fpersico) [reply ts/store/reducers/index.ts]: Not really.

**Comment by cloudify** [comment]
@fpersico what happens to this after https://github.com/teamdigitale/italia-app/pull/498 ?
